core_types/src/cas/storage.rs

```diff
@@ -628,6 +638,71 @@ impl EavTestSuite {
             });
         }
     }
+
+    pub fn test_tombstone<A, S>(mut eav_storage: S)
+    where
+        A: AddressableContent + Clone,
+        S: EntityAttributeValueStorage,
+    {
+        let foo_content = Content::from(RawString::from("foo"));
+        let bar_content = Content::from(RawString::from("bar"));
+
+        let one = A::try_from_content(&foo_content)
+            .expect("could not create AddressableContent from Content");
+        let two = A::try_from_content(&bar_content)
+            .expect("could not create AddressableContent from Content");
+        let attribute = Attribute::LinkTag("c".into(), "c".into());
+        let mut expected_tombstone = BTreeSet::new();
+        let mut expected_tombstone_not_found = BTreeSet::new();
+
+        vec!["a", "b", "c", "d", "e"].iter().for_each(|s| {
+            let eav = EntityAttributeValueIndex::new(
+                &one.address(),
+                &Attribute::LinkTag(String::from(s.clone()), String::from(s.clone())),
+                &two.address(),
+            )
+            .expect("could not create EAV");
+            let expected_eav = eav_storage
+                .add_eavi(&eav)
+                .expect("could not add eav")
+                .expect("Could not get eavi option");
+            if *s == "c" {
+                expected_tombstone.insert(expected_eav);
+            } else if *s == "e" {
+                expected_tombstone_not_found.insert(expected_eav);
+            } else {
+                ()
+            }
+        });
+
+        let expected_attribute = Some(attribute.clone());
+        assert_eq!(
+            expected_tombstone,
+            eav_storage
+                .fetch_eavi(&EaviQuery::new(
+                    Some(one.address()).into(),
+                    None.into(),
+                    Some(two.address()).into(),
+                    IndexFilter::LatestByAttribute,
+                    Some(expected_attribute.into())
+                ))
+                .unwrap()
+        );
+
+        let expected_last_attribute = Some(Attribute::LinkTag("e".into(), "e".into()));
+        assert_eq!(
+            expected_tombstone_not_found,
+            eav_storage
+                .fetch_eavi(&EaviQuery::new(
+                    Some(one.address()).into(),
+                    None.into(),
+                    Some(two.address()).into(),
+                    IndexFilter::LatestByAttribute,
+                    Some(expected_last_attribute.into())
+                ))
+                .unwrap()
+        );
+    }
```

Comment:
I don't understand this test - that is, I don't understand how this tests a tombstone set implementation.

You are only adding links with `Attribute::LinkTag`. I don't see `Attribute::RemoveLink` used at all. Isn't that what should add a link to the tombstone set?

Can you maybe add some comments to help understand what's going on?

Reply:
I will change this but the RemoveLink aspect of this algorithm is irrelevant because it is able to fold on the desired Attribute + Content. There is a can_remove_links that illustrates this but will change this to RemoveLink.

The fold() function in line 55 is essentially the heart of how the tombstone works.

comments are good and will add them!

Reply:
Ok, let me check if I get this right:
1. You have two entries (addressable contents used as E and V ;))
2. You create several links (EAVIs) between them but with differing link type and link tag.
3. For one of those type/tag combinations you also create a RemoveLink EAVI and another new LinkAdd EAVI.
4. Then you query the EAVI storage for anything between those two entries that has the RemoveLink attribute and you assert that you get back exactly that one EAVI item.

Is that what this code is doing? I still don't understand how that tests tombstone semantic..